### ISSUE TITLE
GPU code is slower than CPU

### ISSUE THREAD

**Issue by gpgvl** [PointCloudLibrary/pcl#2871]
I know there's a mailing list for these questions (www.pcl-users.org) but I tried to send it and it's like imposiblle.

## Your Environment

* Operating System and version: Ubuntu 16.04
* Compiler: catkin
* PCL Version: 1.8.0
* GPU: Nvidia GForce 1060

## Context

I have a code where I use the segmentation module of pcl, and it works but it is slow, so I changed it to use the gpu/segmentation module. Since then my code is even slower.

## Expected Behavior

I suppose that it should work faster on the GPU.

## Current Behavior

I have realized that the problem is in the `extract` funtion from `EuclideanClusterExtraction` class, which works 10 times slower when using GPU. I saw this question [unresolved](https://github.com/PointCloudLibrary/pcl/issues/2703), no one answered and I don't know if it is solved, so I want to know if it is a problem with my computer and configuration or if it is not possible to use these functions with GPU support.

## Code to Reproduce

```
// Create the KdTree object for the search method of the extraction
pcl::gpu::Octree::PointCloud clusterCloud;
clusterCloud.upload(cloud_filtered->points);

pcl::gpu::Octree::Ptr octree_device (new pcl::gpu::Octree);
octree_device->setCloud(clusterCloud);
octree_device->build();

// Create the cluster extractor object for the planar model and set all the parameters
std::vector<pcl::PointIndices> cluster_indices;
pcl::gpu::EuclideanClusterExtraction ec;
ec.setClusterTolerance (0.1);
ec.setMinClusterSize (2000);
ec.setMaxClusterSize (250000);
ec.setSearchMethod (octree_device);
ec.setHostCloud (cloud_filtered);
			
ec.extract (cluster_indices);
```

**Comment by taketwo** [comment]
Is this issue substantially different from #2703 that you referenced?

**Comment by gpgvl** [comment]
Sorry, I think that the difference is that the other issue just said that it is slow, I don't even know if that's a question. I want to know if it is possible to run it faster or not, if someone tried it or wants to do it. I give info too about which function gives the problem. I don't know if it is substantially different, but I don't think I will solve it waiting for someone to answer the other issue. Like I said I tried to send my question to users-pcl but the mail is like broken, so I asked here because maybe someone used this module. Thank you.

**Comment by taketwo** [comment]
Unfortunately I personally don't use this module so can not help.

I just want to minimize issue duplication here. I understand that from your side waiting for someone to answer another issue does not feel like getting further in solving your problem. But asking nearly the same and hoping that this time someone will react is also quite hopeless to be honest. But anyway, we can keep both issues around.

**Comment by gpgvl** [comment]
Thank you. Anyway I don't think now that someone will answer, but I needed to try.

**Comment by eBernalV** [comment]
Hello, I am having the same issue, and I think that the problem is because GPU uses octree while CPU uses kdtree, maybe a GPU kdtree implementation will make it faster?

**Comment by gpgvl** [comment]
@eBernalV maybe. As nobody had a solution I reduced the amount of points to make it faster, so I never found the answer. Did you find a kdtree implementation for GPU?

**Comment by kunaltyagi** [comment]
Side question: I mostly deal with large amount of CPU cores so I also don't know practically how data-structure constraints for GPU differ from the CPU (cache, memory-latency, memory-barriers, numa, etc.). Does anyone here have resources for that?

Asking for a trial-side project (working on an integrated tree implementation (kd, ball, quad, oct) because switching API gears is driving me mad)

**Comment by FabianSchuetze** [comment]
The issue should be addressed by PR #4677
